### ISSUE TITLE
Possible bug in EqualObjectGraphs

### ISSUE THREAD

**Issue by achiyae** [mozilla/rhino#1192]
Hello,

I am trying to compare four objects and get the following result:
a != b, b!=c, b!=d 
AND
a == c, a == d, b == c

:point_right: when I write `a != b` I actually mean that `NativeContinuation.equalImplementations(a, b) == false`.

I am aware of @szegedi 's comment in the class documentation of [EqualObjectGraphs.java](https://github.com/mozilla/rhino/blob/master/src/org/mozilla/javascript/EqualObjectGraphs.java), however, I am not certain that this bug is caused due to cyclic data structure.

In [line 43](https://github.com/mozilla/rhino/blob/8fac71800f1bb65500f7c38fb1dcf2a706b47e36/src/org/mozilla/javascript/EqualObjectGraphs.java#L43), a map is defined to remember all known equals objects: `Map<Object, Object> knownEquals = new IdentityHashMap<>();`. A problem may arise if `a` and `b` are detected as equals and added to the map, and then `b` and `c` are detected as equals. In this case, we will end up with a weird inconsistency situation in which `a` equals `b`, and `b` equals `c`; however, `b` does not equals `a` (note: `c` does not equals `a` either, though it's okay because we didn't check it yet...)

The use of `Map<Object, Object>` dictates that an object can be equals to only one object, which of course is not true. Since there is no cyclic data structure here, I believe that this may be a bug rather than a feature...

**Comment by tuchida** [comment]
Could you show us the smallest sample code to reproduce?

**Comment by eggsterino** [comment]
Here:

```javascript
package org.mozilla.javascript;

public class Main {
  public static void main(String[] args) {
    var o1 = makeObject1();
    var o2 = makeObject2();
    System.out.println(new EqualObjectGraphs().equalGraphs(o1, o2)); // prints false
  }

  private static Object makeObject1() {
    var o1 = new Object[]{"3"};
    var o2 = new Object[]{"3"};
    return new Object[]{o1, o2};
  }

  private static Object makeObject2() {
    var o1 = new Object[]{"3"};
    return new Object[]{o1, o1};
  }
}
```

**Comment by szegedi** [comment]
This is by design. The object graphs for these are different. I created a
Grapviz graph for you to illustrate the difference: https://bit.ly/3pSXgHz
This method specifically performs structural comparison, not just value
comparison.

Hope this helps,
  Attila.

**Comment by eggsterino** [comment]
Thanks Attila, it does. 
I understand why this example is not good, however I do have a reproduceable example with four continuations as I presented in the first message. The problem is that it is an extremely complicated example with too many objects and nested functions.
I'll work on a simpler example...

**Comment by szegedi** [comment]
@eggsterino I have thought more about what you said here:

> The use of Map<Object, Object> dictates that an object can be equals to only one object, which of course is not true.

and realized that in this case, it _is_ true. Here's why. This method compares _two_ object graphs for structural equality, which is a mixture of value equality and referential topological equality. Essentially, it needs to establish that there's a bijection "f" between two object graphs G1 and G2 (without being mathematically strict: such that the values of leaves are equivalent and the edges between them are the same.) For such a bijection, you always need object _pairs_: which object from G1 is the pair of the object from G2. Therefore, a `Map<Object, Object>` is sufficient. If graph G1 has two objects that are equivalent by value, but distinct identities, they're different for this purpose, and G2 should also have two similar objects. A very small example would be:
```javascript
[ { a: 1 }, {a: 1 } ]
```
sure, both elements of the array are value-equivalent objects, but they're different objects. If you'd compare this to another object with the same values, the bijection (expressed as edges between objects in G1 and G2) would be:
```javascript
G1: [ { a: 1 }, { a: 1 } ]
         ↑         ↑
         ↓         ↓
G2: [ { a: 1 }, { a: 1 } ]
```
Since the map always expresses mapping of objects from graph G1 to G2 and vice versa, it should always be enough for it to contain pairs of objects; always one from G1 and one from G2. If one object from G1 can be mapped to multiple objects in G2 or the other way round, they are not structurally equivalent.

**Comment by eggsterino** [comment]
Dear @szegedi 

Indeed I can see why what you wrote is correct. Nevertheless, I have these impossible continuations.
I tried to simplify the code to make the example as minimal as possible, but no success so far, so here it is...

To reproduce my bug:
1. Clone the code from [StateSpaceMapper](https://github.com/bThink-BGU/StateSpaceMapper) and switch to branch "bug".
2. Place a breakpoint at [GenerateAllTracesInspection.java#L160](https://github.com/bThink-BGU/StateSpaceMapper/blob/d43015422fc99aa6f837fab1b321753a8de86737/src/main/java/il/ac/bgu/cs/bp/statespacemapper/GenerateAllTracesInspection.java#L160).
3. Run [SpaceMapperRunner](https://github.com/bThink-BGU/StateSpaceMapper/blob/bug/src/main/java/il/ac/bgu/cs/bp/statespacemapper/SpaceMapperRunner.java) in debug mode a few times until you reach the breakpoint (it doesn't happen every time). 

As you will notice, the first equality (line 157) returns false and the following two equalities (158-159) return true. Meanning that:  `a != b && c == a && c == b`.

From what I have found, `a` and `b` are at the same location inside an InterpretedFunction called "sync". During the equality check of these continuations, the prototype object of this InterpretedFunction is being checked for equality. During this equality check, the IDs of these prototypes are being tested and when it tries to check equality for the ID (i.e., string) "constructor" - it returns false because `prev2 != null`
